### PR TITLE
feat: add validation layer for `metadata_properties` and `metadata_filters`

### LINKED ISSUE #3859: [FEATURE] Add metadata validation in `add_records` and `filter_by`
## Description

A validation layer should be included for both the `metadata_properties` and the `metadata_filters` (linked with the pre-defined `metadata_properties`) on both `add_records` and `filter_by` methods, respectively.

This validation will help on detecting the problem on the user-side to avoid calling the API with wrong payloads, and also to help the user debug it better.

### PR DESCRIPTION
**Pull request by alvarobartt**
# Description

This PR adds a validation layer for both the `metadata_properties` when `metadata` is provided via `FeedbackRecord` when adding that record to either a local or existing `FeedbackDataset`, as well as validating the `metadata_filters` provided to the `filter_by` method for `RemoteFeedbackDataset`.

These validation layers would help reduce the API workload as the validation will be done client-side, which implies that the client will handle the validation first.

In this case, as we have the following properties: `TermsMetadataProperty`, `IntegerMetadataProperty` and `FloatMetadataProperty`; we'll need to ensure that the values provided for those are correct, both on `add_records` and also on `filter_by`.

The approach is similar to the one already implemented for the `fields` of a `FeedbackRecord`, but extending it to the `metadata_properties` to also include a validation layer that ensures not just the type of the value provided, but whether it's compliant with the pre-defined values for `TermMetadataProperty` or the pre-defined boundaries for both `IntegerMetadataProperty` and `FloatMetadataProperty`. The core difference is that besides the `pydantic` validation, there's also a method in each property that validates the values provided in the filters related to those metadata properties.

Closes #3859

**Type of change**

- [X] New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

- [x] Add unit tests for the `generate_pydantic_schema_for_metadata` method
- [x] Add unit tests for both `_validate_filter` and `_pydantic_field_with_validator` in each property i.e. `TermsMetadataProperty`, `IntegerMetadataProperty`, and `FloatMetadataProperty`
- [x] Add unit tests for `add_records` with both correct and wrong `metadata`
- [x] Add integration tests for `add_records` and `filter_by` to ensure that the validation is correct both before and after applying those ops

**Checklist**

- [ ] I added relevant documentation
- [X] follows the style guidelines of this project
- [X] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [x] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

### CODE REVIEW

File: src/argilla/client/feedback/schemas/metadata.py [hunk 1]
```diff
@@ -98,6 +103,19 @@ def check_values(cls, terms_values: List[str], values: Dict[str, Any]) -> List[s
     def server_settings(self) -> Dict[str, Any]:
         return {"type": self.type, "values": self.values}
 
+    def _all_values_exist(self, introduced_value: str) -> None:
+        if introduced_value not in self.values:
+            raise ValueError(
+                f"Provided '{self.name}={introduced_value}' is not valid, only values in {self.values} are allowed."
+            )
```
Comment (frascuchon) [inline src/argilla/client/feedback/schemas/metadata.py:126]: Is this a pydantic validator?

Comment (alvarobartt) [reply src/argilla/client/feedback/schemas/metadata.py:126]: Well it's just a class-method in the `pydantic.BaseModel` which is later used as a validator in the property `_pydantic_field_with_validator`

**Comment by alvarobartt** [comment]
FYI @frascuchon I had to revert the `return 0` in `FilteredRemoteRecords.__len__` magic method as returning 0 is misleading IMO, and we cannot return e.g. -1, anyway I assume this is temporarily as before the release we can just add an API call to get the total number of records for filtered datasets.

File: src/argilla/client/feedback/dataset/remote/dataset.py [hunk 2]
```diff
@@ -177,12 +178,36 @@ def filter_by(
...
-        if response_status and not isinstance(response_status, list):
-            response_status = [response_status]
-        if metadata_filters and not isinstance(metadata_filters, list):
-            metadata_filters = [metadata_filters]
 
-        #  accessing records later
+        if response_status:
+            if not isinstance(response_status, list):
+                response_status = [response_status]
+            if not all(status in [arg.value for arg in FeedbackResponseStatusFilter] for status in response_status):
```
Comment (gabrielmbmb) [inline src/argilla/client/feedback/dataset/remote/dataset.py:185]: I would create a constant called `_RESPONSE_STATUS_FILTER_VALUES` or something similar with value `[arg.value for arg in FeedbackResponseStatusFilter]`

File: src/argilla/client/feedback/dataset/remote/dataset.py [hunk 3]
```diff
@@ -177,12 +178,36 @@ def filter_by(
...
+                raise ValueError(
+                    f"Invalid `response_status={response_status}` provided, must be one"
+                    f" of: {[arg.value for arg in FeedbackResponseStatusFilter]}"
+                )
+
+        if metadata_filters:
+            if not isinstance(metadata_filters, list):
+                metadata_filters = [metadata_filters]
+            if not all(
+                metadata_filter.name in [metadata_property.name for metadata_property in self.metadata_properties]
```
Comment (gabrielmbmb) [inline src/argilla/client/feedback/dataset/remote/dataset.py]: I would create a variable `metadata_properties_names` just after the `if metadata_filters`

File: src/argilla/client/feedback/dataset/base.py [hunk 4]
```diff
@@ -213,8 +213,28 @@ def question_by_name(self, name: str) -> Union[AllowedQuestionTypes, "AllowedRem
     def metadata_properties(
         self,
     ) -> Union[List["AllowedMetadataPropertyTypes"], List["AllowedRemoteMetadataPropertyTypes"]]:
+        """Returns the metadata properties that will be indexed and could be used to filter the dataset."""
         return self._metadata_properties
 
+    def metadata_property_by_name(
```
Comment (gabrielmbmb) [inline src/argilla/client/feedback/dataset/base.py:219]: maybe we could cache the return value of this function

**Pull request merged**